### PR TITLE
Add RangeFilter

### PR DESCRIPTION
**Pull request by ahlee2326**
Filter collections from range, can be either (between / lt / lte / gt / gte). This is useful when dealing with min price and max price, duration and etc.

Sample usage: 
dummy?dummyPrice[between]=12.99..19.99
dummy?dummyPrice[gt]=15.99&dummyPrice[lte]=19.99
...

### CODE REVIEW

File: Doctrine/Orm/Filter/RangeFilter.php [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+     */
+    private function addWhere(QueryBuilder $queryBuilder, $alias, $field, $operator, $value)
+    {
+        $valueParameter = QueryNameGenerator::generateParameterName(sprintf('%s_%s', $field, $operator));
+
+        switch ($operator) {
+            case self::PARAMETER_BETWEEN:
+                $rangeValue = explode('..', $value);
+
+                if (empty($rangeValue) || count($rangeValue) > 2) {
```
Comment (sroze) [inline Doctrine/Orm/Filter/RangeFilter.php]: `$rangeValue` will actually never be empty, but `count($rangeValue)` can be 1.

File: Tests/Doctrine/Orm/Filter/RangeFilterTest.php [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+    /**
+     * Providers 3 parameters:
+     *  - filter parameters.
+     *  - properties to test. Keys are the property name. If the value is true, the filter should work on the property,
+     *    otherwise not.
+     *  - expected DQL query.
+     *
+     * @return array
+     */
+    public function filterProvider()
```
Comment (sroze) [inline Tests/Doctrine/Orm/Filter/RangeFilterTest.php]: Could you add a test for `between`? Also with the failure cases?

File: Doctrine/Orm/Filter/RangeFilter.php [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                    $queryBuilder->join(sprintf('%s.%s', $parentAlias, $association), $alias);
+                    $parentAlias = $alias;
+                }
+
+                $field = $propertyParts['field'];
+            }
+
+            foreach ($values as $operator => $value) {
+                if (
+                    $operator === self::PARAMETER_BETWEEN ||
```
Comment (sroze) [inline Doctrine/Orm/Filter/RangeFilter.php]: Maybe you can use `in_array` instead?

**Comment by sroze** [comment]
Thank you this PR @ahlee2326! In addition to the comments I've put, could you create a documentation PR?

**Comment by theofidry** [comment]
Maybe @dunglas changed his stance since then but last time we talked about it we agreed that we wanted to keep the bundle simple and advanced filters could be third-party libraries like [LoopBackApiBundle](https://github.com/theofidry/LoopBackApiBundle).

**Comment by ahlee2326** [comment]
But I think these filters are quite simple and general that most of the application will need. Avoiding user from reinventing the wheel is important.

**Comment by theofidry** [comment]
>But I think these filters are quite simple and general that most of the application will need.

Like having a `neq`, `like` and `nlike`. And then you find yourself 3/4 of the functionalities available in LoopBackApiBundle, which I originally created because the filter provided out of the box with DunglasApiBundle was not advanced enough for my taste but was not about being improved because it would mean having too complex filters.

What I have said is not final, but we must draw a line. We're already providing two filters with a multi-property search (order and where) with support for null ordering. They alone takes already over 1300 LoC out of the 8000 LoC of the filters. That's not a trivial amount for something that is not the core of the library but rather an extension, and specific to Doctrine ORM.

That's why I would rather keep the core library simple and have another library for a more advanced filters.

**Comment by sroze** [comment]
I would actually even remove all this Doctrine stuff in an extension as our goal is to be able to be completely decoupled from it. Anyway, before that move I give a :+1: for this feature to be merged (after the review process is finished obviously)

**Comment by dunglas** [comment]
What's the problem with the current design @sroze? You can install the bundle without the Doctrine dependency.

I'm for integrating this filter in the bundle. We can also integrate the loopback if it is stable!

**Comment by theofidry** [comment]
I think @sroze meant that he would rather see Doctrine related stuff in another repo or something and not mixing it with the library.

Concerning integrating the filter thing see the gitter chat to avoid polluting this thread

**Comment by dunglas** [comment]
What's the interest of doing that? Decoupling things is important but having a single repository is easier for maintenance and tests.

Even if everything is in a same repository, you can install the bundle without Doctrine, Pomm...

Pomm support is in progress and will be in the same repository too.

**Comment by sroze** [comment]
@dunglas as I said I'm in favor of merging this PR :wink:

File: Doctrine/Orm/Filter/RangeFilter.php [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                    $alias = QueryNameGenerator::generateJoinAlias($association);
+                    $queryBuilder->join(sprintf('%s.%s', $parentAlias, $association), $alias);
+                    $parentAlias = $alias;
+                }
+
+                $field = $propertyParts['field'];
+            }
+
+            foreach ($values as $operator => $value) {
+                if (in_array($operator, [
```
Comment (dunglas) [inline Doctrine/Orm/Filter/RangeFilter.php]: `in_array` is slow. Can you use a switch instead?

```php
switch ($operator) {
    case self::PARAMETER_BETWEEN:
    case self::PARAMETER_GREATER_THAN:
   // ...
}```

File: Doctrine/Orm/Filter/RangeFilter.php [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+     */
+    private function addWhere(QueryBuilder $queryBuilder, $alias, $field, $operator, $value)
+    {
+        $valueParameter = QueryNameGenerator::generateParameterName(sprintf('%s_%s', $field, $operator));
+
+        switch ($operator) {
+            case self::PARAMETER_BETWEEN:
+                $rangeValue = explode('..', $value);
+
+                if (count($rangeValue) !== 2) {
```
Comment (dunglas) [inline Doctrine/Orm/Filter/RangeFilter.php]: We use Yoda style (Symfony CS): `2 !== count($rangeValue)`

**Comment by dunglas** [comment]
Good job @ahlee2326. I left some minor comments to fix before the merge. Can you also add a doc in https://github.com/api-platform/doc please?

Thanks.

**Comment by ahlee2326** [comment]
@dunglas Updated. I will update the doc later.

**Comment by dunglas** [comment]
:+1:

File: Tests/Doctrine/Orm/Filter/RangeFilterTest.php (hunk 2 above)
Comment (sroze) [inline Tests/Doctrine/Orm/Filter/RangeFilterTest.php]: Fixed.

**Comment by sroze** [comment]
Nice, thank you very much @ahlee2326 ! 
I'll issue a PR regarding to the CS after as it's not your changes only (or at all).

**Pull request merged**